enable_retired_username_email/signals.py

```diff
@@ -0,0 +1,72 @@
+"""
+Signal handler for setting setting modified username and email hash in user model.
+"""
+
+import time
+
+from django.contrib.auth.models import User
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+
+from enable_retired_username_email.compat import get_retired_email_hash, get_user_retirement_status_model
+
+UserRetirementStatus = get_user_retirement_status_model()
+
+
+@receiver(post_save, sender=UserRetirementStatus)
+def replace_username_email_hash(
+    instance: UserRetirementStatus,
+    created,
+    **kwargs
+):  # pylint: disable=unused-argument
+    """
+    This function is triggered when UserRetirementStatus is saved.
+
+    Upon receiving the post_save
+    signal for the UserRetirementStatus model,
+    appends to the original username and email and
+    calculates their hash values. These are then
+    set to the User model and the corresponding
+    UserRetirementStatus is deleted
+    """
+    if created:
+        instance.retired_username = get_new_retired_username(instance.original_username)
+        instance.save()
+
+    elif instance.current_state.state_name == 'COMPLETE':
+        modified_email_hash = get_modified_email_hash(instance)
+
+        user = User.objects.get(username=instance.retired_username)
+        user.email = modified_email_hash
+        user.save()
```

Comment:
We can just move this to the above block right?

Reply:
Do you mean moving the block into the `get_modified_email_hash()` function ?

If so, I like one of our comments below which suggests modifying `get_modified_email_hash()` to have `str` argument and `str` output.

I think if we keep this block here, the broad intention of the `else` block is clearer at a glance. The reader only has to go into the `get_modified_email_hash` if they want to get the details of how modified hash is generated.

Let me know if you feel otherwise.

Reply:
What I am trying to say is why do we need to wait for the status to be completed to replace the email with a hashed one, we should be doing it when we replace the username, right?

When the status is complete we can easily delete the entry :)

Reply:
There are two reasons for this :

1. The primary reason : 
     There are 3 entrypoints for initiating retirement of user account. In all three places : [gdpr_user_retirement](https://github.com/open-craft/edx-platform/blob/03106b60d04ac50a24d116acfdcc1bc5824e27b6/lms/djangoapps/gdpr_user_retirement/views.py#L62), [DeactivateLogoutView](https://github.com/open-craft/edx-platform/blob/03106b60d04ac50a24d116acfdcc1bc5824e27b6/openedx/core/djangoapps/user_api/accounts/views.py#L442) and [retire_user command](https://github.com/open-craft/edx-platform/blob/03106b60d04ac50a24d116acfdcc1bc5824e27b6/openedx/core/djangoapps/user_api/management/commands/retire_user.py#L75), the email hash is calculated again after entry in `UserRetirementStatus` model is created and the `user` model is saved with the email hash. In our plugin, setting the modified email hash in the first block, will create a race condition, and will be potentially overridden with the original hash value.

2. The secondary reason:
     The email id is the primary identifier used to refer the `user` model and `UserRetirementStatus` model throughout the retirement flow. Thus we would require thorough analysis and testing to ensure no flow is impacted if we replace the email at the start of the retirement flow.

Reply:
The first reason sounds good to me :)